### PR TITLE
Quick way to apply Chia BLS

### PR DESCRIPTION
**Pull request by ChihChengLiang**
### What was wrong?

Before properly addressing #746, this is a minimal way to get Chia BLS after #714 is merged.

### How was it fixed?

- Add BaseBLSBackend and ChiaBackend
- Fix invalid private key issue

#### Cute Animal Picture

image

### CODE REVIEW

**Comment by ChihChengLiang** [comment]
Some odd bugs summarized at #767

File: tests/eth2/bls-bindings/chia_network/test_api.py [hunk 1]
```diff
@@ -67,23 +75,38 @@ def test_sanity():
...
+        with pytest.raises(ValidationError):
+            sig = sign(msg, privkey, domain=domain)
+
+        with pytest.raises(ValidationError):
+            pub = privtopub(privkey)
+
+
+def test_empty_aggregation():
+    assert aggregate_pubkeys([]) == EMPTY_PUBKEY
+    assert aggregate_signatures([]) == EMPTY_SIGNATURE
```
Comment (ChihChengLiang) [inline tests/eth2/bls-bindings/chia_network/test_api.py]: Note py_ecc has this behavior.

File: tests/eth2/core/beacon/tools/builder/test_builder_validator.py [hunk 2]
```diff
@@ -79,7 +82,11 @@ def test_aggregate_votes(votes_count, random, privkeys, pubkeys):
     assert len(voted_index) == len(votes)
 
     aggregated_pubs = bls.aggregate_pubkeys(pubs)
-    assert bls.verify(message_hash, aggregated_pubs, sigs, domain)
```
Comment (ChihChengLiang) [inline tests/eth2/core/beacon/tools/builder/test_builder_validator.py]: To py_ecc, if you sign a message with empty public key and empty signature, it verifies True. That seems wrong.
```python
In [10]: from py_ecc.bls import verify

In [11]: verify(b'\x56'*32, b'\xc0'+ b'\x00'*47, b'\xc0'+ b'\x00'*95, 0)
Out[11]: True
```

Comment (ChihChengLiang) [reply tests/eth2/core/beacon/tools/builder/test_builder_validator.py]: Created an issue here https://github.com/ethereum/py_ecc/issues/74

File: eth2/_utils/bls_bindings/chia_network/api.py [hunk 3]
```diff
@@ -14,9 +14,46 @@
...
-    return privkey.to_bytes(bls_chia.PrivateKey.PRIVATE_KEY_SIZE, "big")
+
+def _pubkey_from_bytes(pubkey: BLSPubkey) -> 'bls_chia.PublicKey':
+    try:
+        return bls_chia.PublicKey.from_bytes(pubkey)
+    except (RuntimeError, ValueError) as error:
+        raise ValidationError(f"Bad public key: {pubkey}, {error}")
+
+
+def _signature_from_bytes(signature: BLSSignature) -> 'bls_chia.Signature':
```
Comment (mhchia) [inline eth2/_utils/bls_bindings/chia_network/api.py]: ditto

File: eth2/_utils/bls_bindings/chia_network/api.py [hunk 4]
```diff
@@ -52,6 +89,9 @@ def verify(message_hash: Hash32, pubkey: BLSPubkey, signature: BLSSignature, dom
 
 
 def aggregate_signatures(signatures: Sequence[BLSSignature]) -> BLSSignature:
+    if len(signatures) == 0:
+        return EMPTY_SIGNATURE
```
[the remark this answers is not in the document]
Comment (ChihChengLiang) [reply eth2/_utils/bls_bindings/chia_network/api.py]: Yes, see https://github.com/ethereum/eth2.0-specs/blob/dev/specs/bls_signature.md#bls_aggregate_pubkeys
> When len(pubkeys) == 0 the empty sum is the G1 point at infinity.
> When len(signatures) == 0 the empty sum is the G2 point at infinity.

**Comment by ChihChengLiang** [comment]
Plan to rebase and merge this after #714. And will request another round of review after adding these features to this PR.

- Some tests for empty aggregation
- Raise Validation error for failing validation instead of simply returning False.

**Comment by ChihChengLiang** [comment]
close in favor of #787

**Pull request closed without merging**